pkg/kncloudevents/message_sender_test.go

```diff
@@ -194,6 +196,345 @@ func TestHTTPMessageSenderSendWithRetriesWithSingleRequestTimeout(t *testing.T)
+			config: &RetryConfig{
+				RetryMax:              2,
+				RetryAfterMaxDuration: &largeRetryAfterMaxDuration,
+			},
+			statusCode:     http.StatusInternalServerError,
+			responseFormat: None,
+			wantReqCount:   3,
+		},
+
+		//
+		// Small Max Tests (Less Than Retry-After Value)
+		//
+
+		{
+			name: "small max 429 without Retry-After",
+			config: &RetryConfig{
+				RetryMax:              2,
+				RetryAfterMaxDuration: &smallRetryAfterMaxDuration,
+			},
+			statusCode:     http.StatusTooManyRequests,
+			responseFormat: None,
+			wantReqCount:   3,
+		},
+		{
+			name: "small max 429 with Retry-After seconds",
+			config: &RetryConfig{
+				RetryMax:              2,
+				RetryAfterMaxDuration: &smallRetryAfterMaxDuration,
+			},
+			statusCode:      http.StatusTooManyRequests,
+			responseFormat:  Seconds,
+			responseSeconds: 4,
+			wantReqCount:    3,
+		},
+		{
+			name: "small max 429 with Retry-After date",
+			config: &RetryConfig{
+				RetryMax:              2,
+				RetryAfterMaxDuration: &smallRetryAfterMaxDuration,
+			},
+			statusCode:      http.StatusTooManyRequests,
+			responseFormat:  Date,
+			responseSeconds: 2,
+			wantReqCount:    3,
+		},
+		{
+			name: "small max 429 with invalid Retry-After",
+			config: &RetryConfig{
+				RetryMax:              2,
+				RetryAfterMaxDuration: &smallRetryAfterMaxDuration,
+			},
+			statusCode:     http.StatusTooManyRequests,
+			responseFormat: Invalid,
+			wantReqCount:   3,
+		},
+		{
+			name: "small max 503 with Retry-After seconds",
+			config: &RetryConfig{
+				RetryMax:              2,
+				RetryAfterMaxDuration: &smallRetryAfterMaxDuration,
+			},
+			statusCode:      http.StatusServiceUnavailable,
+			responseFormat:  Seconds,
+			responseSeconds: 4,
+			wantReqCount:    3,
+		},
+		{
+			name: "small max 500 without Retry-After",
+			config: &RetryConfig{
+				RetryMax:              2,
+				RetryAfterMaxDuration: &smallRetryAfterMaxDuration,
+			},
+			statusCode:     http.StatusInternalServerError,
+			responseFormat: None,
+			wantReqCount:   3,
+		},
+	}
+
+	// Execute The TestCases
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+
+			// Consistent CheckRetry & Backoff Implementation For All TestCases
+			paddingDuration := 250 * time.Millisecond // Test Execution Allowance For Resending Requests
+			backoffDuration := 500 * time.Millisecond // Constant Backoff Time For Simplified Testing
+			tc.config.CheckRetry = SelectiveRetry
+			tc.config.Backoff = func(attemptNum int, resp *http.Response) time.Duration { return backoffDuration }
+
+			// Determine The Response Retry-After Backoff Duration From TestCase
+			var responseDuration time.Duration
+			switch tc.responseFormat {
+			case None, Invalid:
+			case Seconds, Date:
+				if tc.responseSeconds > 0 {
+					responseDuration = time.Duration(tc.responseSeconds) * time.Second
+				}
+			default:
+				assert.Fail(t, "TestCase with unsupported ResponseFormat '%v'", tc.responseFormat)
+			}
```

Comment:
I'd prefer removing logic in the test and having values specified without deriving them on the fly.

Reply:
Yeah, good call - cleaner that way - will fix!